### DISCUSSION TITLE
Timeseries: Finding the first tag to reach a specific value?

### DISCUSSION THREAD

**Discussion by trevhunter** [ravendb/ravendb#14820] [Q&A]
Imagine I have a race with two cars (`Car1` and `Car2`). The winner of the race is the car that has done 1000km first. Each car's odometer sends its current value to RavenDB every 10 seconds. 

How would I model the timeseries and timeseries query to figure out which car won the race?

Option 1: Model the race as a single timeseries, with each car having a different tag attached to each odometer value. Query the timeseries to find the earliest date any value has crossed 1000km and get the tag of that value. 

Option 2: Model each car's odometer as two different documents/timeseries and somehow combine both timeseries in a query.

**Comment by karmeli87** [comment]
Another option will be to model each race as a document and if you are expecting only 2 cars you can put them in the _same_ timeseries (one timeseries supports upto 32 values) and query for the last result. 

Would that work for you case?

**Comment by ayende** [comment]
Each car would probably be a separate document, and you have a times series `odometer` which holds the values for that.
That is the best model option, I think.

Now, for queries, the problem we have is that your query doesn't match the manner of operations that we run on time series.
Time series queries operate on the time dimension first, so to ask "first value bigger than X" means that we'll have to scan the entire time series. 
Not advisable if you have a lot of values. 

You can do this via an index, where you index all the time series values above the threshold and find the minimum time.
That may require a storage usage if you have a lot of values, but you'll get the results instantly.

Alternatively, you can do divide & conquer on the server side, using something like this:

```
$start = "2020-05-01T00:00:00.0000000Z"
$end  = "2020-05-30T00:00:00.0000000Z"
$threshold = 80

declare timeseries segmentMonth(e, start, end){
    from e.HeartRates 
    between start and end 
    group by '1 month'
    select last()
}
declare timeseries segmentHour(e, start, end){
    from e.HeartRates 
    between start and end 
    group by '1 hour'
    select last()
}
declare function project(e) {
    var results = segmentMonth(e, $start, $end)
    for(var i = 0; i < results.Results.length; i ++){
        if(results.Results[i].Last[0] >  $threshold)
        {
            var hourly = segmentHour(e, results.Results[i].From, results.Results[i].To)
            for(var j = 0; j < hourly.Results.length; j ++){
                if(hourly.Results[j].Last[0] >  $threshold)
                {
                    return {
                        Results: hourly.Results[j],
                    };
                }
            }
        }
    }
    return {Results:[]}; // no results
}
from "Employees" as e
where id(e) in ('employees/1-A')
select project(e)
```

Warning, rough code above.

Can you provide more details about your scenario?
